File: 242_valid_anagram/isAnagramHash.go
func isAnagram(s string, t string) bool {
	n, m := len(s), len(t)

	if n != m {
		return false
	}

	hash := map[rune]int{}

	for _, c := range s {
		val := hash[c]
		hash[c] = val + 1
	}

	for _, c := range t {
		val := hash[c]
		hash[c] = val - 1

		if hash[c] < 0 {
			return false
		}
	}

	return true
}

File: 242_valid_anagram/isAnagramHash.ts
function isAnagram(s: string, t: string): boolean {
  const n = s.length;
  const m = t.length;

  if (n !== m) return false;

  const hash = new Map<string, number>();

  for (const i of s) {
    const value = hash.get(i) || 0;
    hash.set(i, value + 1);
  }

  for (const i of t) {
    const value = hash.get(i) || 0;
    hash.set(i, value - 1);

    if (hash.has(i) && hash.get(i) < 0) {
      return false;
    }
  }

  return true;
}
